[SaferCPP] Unchecked call arguments in HTMLMediaElement

atar13 opened the pull request:
#### d3e95cb9a1e9e701bf8c074785e4c55e02353f00
<pre>
[SaferCPP] Unchecked call arguments in HTMLMediaElement
<a href="https://bugs.webkit.org/show_bug.cgi?id=301107">https://bugs.webkit.org/show_bug.cgi?id=301107</a>
<a href="https://rdar.apple.com/163052480">rdar://163052480</a>

Reviewed by Ryosuke Niwa.

Change call arguments in HTMLMediaElement.cpp to be checked.
Most of these changes are using protectedDocument() instead
of document(). The rest add usage of Ref/RefPtr/CheckedRef.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::pauseAfterDetachedTask):
(WebCore::HTMLMediaElement::load):
(WebCore::HTMLMediaElement::loadResource):
(WebCore::HTMLMediaElement::isSafeToLoadURL const):
(WebCore::HTMLMediaElement::dispatchPlayPauseEventsIfNeedsQuirks):
(WebCore::HTMLMediaElement::updateShouldContinueAfterNeedKey):
(WebCore::HTMLMediaElement::mediaPlayerCachedKeyForKeyId const):
(WebCore::HTMLMediaElement::mediaPlayerKeyNeeded):
(WebCore::HTMLMediaElement::mediaPlayerMediaKeysStorageDirectory const):
(WebCore::HTMLMediaElement::mediaPlayerInitializationDataEncountered):
(WebCore::HTMLMediaElement::setAudioOutputDevice):
(WebCore::HTMLMediaElement::seekTask):
(WebCore::HTMLMediaElement::finishSeek):
(WebCore::HTMLMediaElement::controls const):
(WebCore::HTMLMediaElement::setVolume):
(WebCore::HTMLMediaElement::mediaPlayerDidReportGPUMemoryFootprint):
(WebCore::HTMLMediaElement::selectNextSourceChild):
(WebCore::HTMLMediaElement::updatePlayState):
(WebCore::HTMLMediaElement::resume):
(WebCore::HTMLMediaElement::elementIsHidden const):
(WebCore::HTMLMediaElement::exitFullscreen):
(WebCore::HTMLMediaElement::updateSleepDisabling):
(WebCore::HTMLMediaElement::mediaPlayerContentBoxRect const):
(WebCore::HTMLMediaElement::mediaPlayerCachedResourceLoader):
(WebCore::HTMLMediaElement::mediaPlayerCreateResourceLoader):
(WebCore::HTMLMediaElement::mediaPlayerEngineFailedToLoad):
(WebCore::HTMLMediaElement::removeBehaviorRestrictionsAfterFirstUserGesture):
(WebCore::HTMLMediaElement::setControllerJSProperty):
(WebCore::HTMLMediaElement::ensureMediaControls):
(WebCore::HTMLMediaElement::supportsSeeking const):
(WebCore::HTMLMediaElement::processingUserGestureForMedia const):
(WebCore::HTMLMediaElement::documentSecurityOrigin const):
(WebCore::HTMLMediaElement::updateMediaPlayer):
(WebCore::HTMLMediaElement::setShowingStats):
(WebCore::HTMLMediaElement::logTextTrackDiagnostics):
(WebCore::HTMLMediaElement::watchtimeTimerFired):
(WebCore::HTMLMediaElement::invalidateBufferingStopwatch):
(WebCore::HTMLMediaElement::limitedMatroskaSupportEnabled const):

Canonical link: <a href="https://commits.webkit.org/302006@main">https://commits.webkit.org/302006@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b6021343417b6a9642097ea0d018987e599b1fd4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127180 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46816 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37956 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134244 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78735 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/dca51ac8-0392-4959-8ea3-e8e6f31fc92a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129052 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47432 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55341 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96789 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64843 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/1a87bf16-11ef-4503-95e3-08234823b856) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130128 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37968 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113909 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77293 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c259ffac-b8e4-4e15-8528-7c9edaac2d81) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36871 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32043 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77624 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107845 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32391 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136727 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53833 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41479 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105308 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54341 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110261 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104996 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26883 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50519 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28964 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/51402 "The change is no longer eligible for processing.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53776 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59863 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/53008 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56341 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54769 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

atar13 commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -3395,14 +3395,19 @@ void HTMLMediaElement::updateShouldContinueAfterNeedKey()
+
+RefPtr<WebKitMediaKeys> HTMLMediaElement::protectedWebkitKeys() const
+{
+    return m_webKitMediaKeys;
```
For some reason the compiler didn't let me put this function in the header. Something about a missing forward declaration for `WebKitMediaKeys`.

atar13 replied:
Ended up removing this function.

rniwa commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -3395,14 +3395,19 @@ void HTMLMediaElement::updateShouldContinueAfterNeedKey()
 RefPtr<ArrayBuffer> HTMLMediaElement::mediaPlayerCachedKeyForKeyId(const String& keyId) const
 {
-    return m_webKitMediaKeys ? m_webKitMediaKeys->cachedKeyForKeyId(keyId) : nullptr;
+    return protectedWebkitKeys() ? protectedWebkitKeys()->cachedKeyForKeyId(keyId) : nullptr;
```
Instead of calling `protectedWebkitKeys()` twice, Store `m_webKitMediaKeys` in a `RefPtr` local variable and call cachedKeyForKeyId on it if it's not null.

atar13 replied:
I removed the `protectedWebkitKeys()` method and just made a `RefPtr` local variable since it's only used here. Let me know what you think.

This is how it looks now https://github.com/WebKit/WebKit/pull/52665/commits/b6021343417b6a9642097ea0d018987e599b1fd4#diff-d57d5e5c415e1c8520a5553a39b00ce95a35369bbb3b5a89e979f463b60ea6b6R3405-R3406

rniwa replied:
Looks good to me.

rniwa commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -8394,7 +8399,8 @@ LayoutRect HTMLMediaElement::mediaPlayerContentBoxRect() const
     if (!renderer)
         return { };
-    return renderer->enclosingBox().contentBoxRect();
+    auto protectedEnclosingBox = CheckedRef { renderer->enclosingBox() };
```
Do this instead:
```cpp
CheckedRef protectedEnclosingBox { renderer->enclosingBox() };
```

atar13 replied:
Changed [here](https://github.com/WebKit/WebKit/pull/52665/commits/b6021343417b6a9642097ea0d018987e599b1fd4#diff-d57d5e5c415e1c8520a5553a39b00ce95a35369bbb3b5a89e979f463b60ea6b6R8398-R8399)

atar13 committed:
[SaferCPP] Unchecked call arguments in HTMLMediaElement
https://bugs.webkit.org/show_bug.cgi?id=301107
rdar://163052480

Reviewed by Ryosuke Niwa.

Change call arguments in HTMLMediaElement.cpp to be checked.
Most of these changes are using protectedDocument() instead
of document(). The rest add usage of Ref/RefPtr/CheckedRef.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::pauseAfterDetachedTask):
(WebCore::HTMLMediaElement::load):
(WebCore::HTMLMediaElement::loadResource):
(WebCore::HTMLMediaElement::isSafeToLoadURL const):
(WebCore::HTMLMediaElement::dispatchPlayPauseEventsIfNeedsQuirks):
(WebCore::HTMLMediaElement::updateShouldContinueAfterNeedKey):
(WebCore::HTMLMediaElement::mediaPlayerCachedKeyForKeyId const):
(WebCore::HTMLMediaElement::mediaPlayerKeyNeeded):
(WebCore::HTMLMediaElement::mediaPlayerMediaKeysStorageDirectory const):
(WebCore::HTMLMediaElement::mediaPlayerInitializationDataEncountered):
(WebCore::HTMLMediaElement::setAudioOutputDevice):
(WebCore::HTMLMediaElement::seekTask):
(WebCore::HTMLMediaElement::finishSeek):
(WebCore::HTMLMediaElement::controls const):
(WebCore::HTMLMediaElement::setVolume):
(WebCore::HTMLMediaElement::mediaPlayerDidReportGPUMemoryFootprint):
(WebCore::HTMLMediaElement::selectNextSourceChild):
(WebCore::HTMLMediaElement::updatePlayState):
(WebCore::HTMLMediaElement::resume):
(WebCore::HTMLMediaElement::elementIsHidden const):
(WebCore::HTMLMediaElement::exitFullscreen):
(WebCore::HTMLMediaElement::updateSleepDisabling):
(WebCore::HTMLMediaElement::mediaPlayerContentBoxRect const):
(WebCore::HTMLMediaElement::mediaPlayerCachedResourceLoader):
(WebCore::HTMLMediaElement::mediaPlayerCreateResourceLoader):
(WebCore::HTMLMediaElement::mediaPlayerEngineFailedToLoad):
(WebCore::HTMLMediaElement::removeBehaviorRestrictionsAfterFirstUserGesture):
(WebCore::HTMLMediaElement::setControllerJSProperty):
(WebCore::HTMLMediaElement::ensureMediaControls):
(WebCore::HTMLMediaElement::supportsSeeking const):
(WebCore::HTMLMediaElement::processingUserGestureForMedia const):
(WebCore::HTMLMediaElement::documentSecurityOrigin const):
(WebCore::HTMLMediaElement::updateMediaPlayer):
(WebCore::HTMLMediaElement::setShowingStats):
(WebCore::HTMLMediaElement::logTextTrackDiagnostics):
(WebCore::HTMLMediaElement::watchtimeTimerFired):
(WebCore::HTMLMediaElement::invalidateBufferingStopwatch):
(WebCore::HTMLMediaElement::limitedMatroskaSupportEnabled const):

Canonical link: https://commits.webkit.org/302006@main
